igvmmeasure/src/igvm_measure.rs

```diff
@@ -0,0 +1,333 @@
+                            &[],
+                        )?;
+                    }
+                }
+                IgvmDirectiveHeader::SnpVpContext {
+                    gpa,
+                    compatibility_mask,
+                    vp_index: _vp_index,
+                    vmsa,
+                } => {
+                    self.measure_vmsa(*gpa, *compatibility_mask, vmsa)?;
+                }
+                _ => (),
+            }
+        }
+        self.log_page(SnpPageType::None, 0, 0);
+
+        Ok(self.digest)
+    }
+
+    fn log_page(&mut self, page_type: SnpPageType, gpa: u64, len: u64) {
+        if self.options.verbose {
+            if (page_type != self.last_page_type) || (gpa != self.last_next_gpa) {
+                if self.last_len > 0 {
+                    println!(
+                        "gpa {:#x} len {:#x} ({} page)",
+                        self.last_gpa, self.last_len, self.last_page_type
+                    );
+                }
+                self.last_len = len;
+                self.last_gpa = gpa;
+                self.last_next_gpa = gpa + len;
+                self.last_page_type = page_type;
+            } else {
+                self.last_len += len;
+                self.last_next_gpa += len;
+            }
+        }
+    }
+
+    fn measure_page(
+        &mut self,
+        gpa: u64,
+        flags: &IgvmPageDataFlags,
+        data_type: IgvmPageDataType,
+        data: &[u8],
+    ) -> Result<(), Box<dyn Error>> {
+        let page_len = if flags.is_2mb_page() {
+            PAGE_SIZE_2M
+        } else {
+            PAGE_SIZE_4K
+        };
+        assert!(data.is_empty() || data.len() == page_len as usize);
+
+        if data.is_empty() {
+            for page_offset in (0..page_len).step_by(PAGE_SIZE_4K as usize) {
+                self.measure_page_4k(gpa + page_offset, flags, data_type, &vec![]);
+            }
+        } else {
+            for (index, page_data) in data.chunks(PAGE_SIZE_4K as usize).enumerate() {
+                self.measure_page_4k(
+                    gpa + index as u64 * PAGE_SIZE_4K,
+                    flags,
+                    data_type,
+                    &page_data.to_vec(),
+                );
+            }
+        }
+        Ok(())
+    }
+
+    fn measure_page_4k(
+        &mut self,
+        gpa: u64,
+        flags: &IgvmPageDataFlags,
+        data_type: IgvmPageDataType,
+        data: &Vec<u8>,
+    ) {
+        let page_info = match data_type {
+            IgvmPageDataType::NORMAL => {
+                if flags.unmeasured() {
+                    self.log_page(SnpPageType::Unmeasured, gpa, PAGE_SIZE_4K);
+                    Some(PageInfo::new_unmeasured_page(self.digest, gpa))
+                } else if data.is_empty() {
+                    self.log_page(SnpPageType::Zero, gpa, PAGE_SIZE_4K);
+                    Some(PageInfo::new_zero_page(self.digest, gpa))
+                } else {
```

Comment:
this is not a legal transformation, we need to insert this as an actual page of zeros:
```suggestion
                } else if data.is_empty() {
                    self.log_page(SnpPageType::Normal, gpa, PAGE_SIZE_4K);
                    Some(PageInfo::new_normal_page(self.digest, gpa, &self.zero_page))
                } else {
```
and then you would cache `zero_page: vec![0u8; PAGE_SIZE_4K as usize]` somewhere.

Reply:
The IGVM loader in QEMU performs the exact same transformation, converting empty pages to `SnpPageType::Zero` so in this case the measurement performed by the firmware is updated in accordance with the same rules as the measure tool and we get matching measurements for the COCONUT IGVM files and QEMU.

The SNP Firmware ABI specification describes the behaviour of the firmware for `PAGE_TYPE_ZERO` as equivalent to inserting a page of zeros - the resulting guest memory is the same, just the measurement is affected. I would assume that there is a performance benefit in using `PAGE_TYPE_ZERO` resulting in faster guest startup but I'm unsure.

So this transformation is legal, its just that the virtualisation stack you are using is interpreting IGVM directives differently to QEMU.

Reply:
I get your point that the outcome is equivalent - but is it worth diverging? The igvm file does say "PAGE_TYPE_NORMAL" so optimizing that into a "PAGE_TYPE_ZERO" feels like a loose interpretation of the file format. I don't know if TDX supports the equivalent of "PAGE_TYPE_ZERO" either.

At this stage we could totally still update Qemu igvm support to match what cloud-hypervisor and other igvm stacks do, it won't be possible to change later.

Reply:
I was following the IGVM specification itself, documented [here under the `file_offset` field](https://docs.rs/igvm_defs/0.1.6/igvm_defs/struct.IGVM_VHS_PAGE_DATA.html#fields). 

> If the file offset is zero, then a page of zeroes is to be loaded. If the page contents are to be measured, and the underlying platform supports measured zero pages as a native page type, then the loader must request this page type; otherwise, the loader must generate a page of zeroes to load as the page contents.

Reply:
Yikes, you're right. Looks like cloud-hypervisor got it wrong when they copied the mshv stack. That ship may have sailed though, what do you think @msft-jlange?

Reply:
Roy's interpretation of the spec is definitely what the spec was intended to do.  However, it appears that other IGVM loaders do not follow that interpretation of the spec.  For example, the Hyper-V loader will supply 4 KB of zeroes with PAGE_TYPE_NORMAL when it sees a zero file offset.  Since this behavior is already in production, it is not practical to declare that the production behavior is buggy because it will cause incompatibilities with IGVM files already in production.  I will plan to change the IGVM spec and I will introduce some sort of new flag information to indicate when a page of zeroes *must* be supplied as PAGE_TYPE_ZERO.

Reply:
In anticipation of the changes to the IGVM specification, I've added an option to the `igvmmeasure` tool to allow the behaviour of handling of zero pages to be changed via a command-line option. This now defaults to using the normal page type rather than the zero page type.

The QEMU branch currently used by COCONUT uses the native zero page type so `Makefile` sets this new option to measure using zero pages. I'll update QEMU then raise a new PR to remove the option (or modify this one if it has not yet merged).